enhancements/ingress/haproxy-router-compression.md

```diff
@@ -0,0 +1,421 @@
+disables the limit. Setting a lower value will prevent the compression work from slowing the whole
+process down and from introducing high latencies.
+
+**tune.comp.maxlevel**: sets the maximum compression level. The compression level affects CPU usage during compression.
+Each session using compression initializes the compression algorithm with this value. The default is 1.
+
+### Open Questions
+### Test Plan
+- Unit tests will validate that HTTP compression directives appear in the `frontend`/`backend` sections of the HAProxy
+  template when it is properly configured in a route.
+- Unit tests will validate that the HTTP compression directive appears in the `defaults` section of the HAProxy template
+  when it is properly configured on the IngressController spec.
+- E2e tests will validate that the IngressController spec accepts and enables compression in the `defaults` section of
+  the generated haproxy.cfg
+
+Note that HAProxy Enterprise **disables** HTTP compression when:
+- The request does not advertise a supported compression algorithm in the "Accept-Encoding" header
+- The response message is not HTTP/1.1 or above
+- The HTTP status code is not one of 200, 201, 202 or 203
+- The response contains neither a "Content-Length" header, nor a "Transfer-Encoding" header whose last value is `chunked`
+- The response contains a "Content-Type" header whose first value starts with `multipart`
+- The response contains a "Cache-Control" header with a value of `no-transform`
+- The response contains a "Content-Encoding" header, indicating that the response is already compressed (see `compression offload`)
+- The response contains an invalid "ETag" header or multiple "Etag" headers
+
+### Graduation Criteria
+
+This feature will be immediately available in release 4.10, and will not have a dev or tech preview.
+
+#### Dev Preview -> Tech Preview
+n/a
+
+#### Tech Preview -> GA
+n/a
+
+#### Removing a deprecated feature
+
+- Announce deprecation and support policy of the existing feature
+- Deprecate the feature
+
+### Upgrade / Downgrade Strategy
+
+An upgrade from a previous installation that had the compression directives exposed, will need to change to use the
+IngressController spec.  No previous installation would have any compression directives via route annotations.
+
+If a cluster administrator applied any of the compression configuration options
+in 4.10, then downgraded to 4.9, the compression configuration settings would no longer be in effect unless they
+used a custom unmanaged HAProxy configuration.
+
+### Version Skew Strategy
+n/a
+
+## Implementation History
+n/a
+
+## Drawbacks
+
+By implementing this proposal we are giving the user tools that could threaten the operation of HAProxy.
+Improper configuration of the MIME types or general over-use of compression could allow the HAProxy to
+dedicate more processing time to compression than to its main duties of routing and traffic management.
+
+## Alternatives
```

Comment:
Because this feature is not security related and because the side-effects are minimal, short-lived, and fixable by the server enforcement by ingress is ok.  

Does HAProxy allow configuration of a default that applies by domain?  If so, is that more natural for customers to configure?  An answer down here in alternatives would help future us remember why the choice was made.

Reply:
Based on feedback from the team, we will not add route-based HTTP compression options in this release.  HAProxy does not allow configuration that applies by domain.